File: app/controllers/Application.java
package controllers;

import static akka.pattern.Patterns.ask;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.stream.Collectors;

import javax.inject.Inject;
import javax.inject.Singleton;

import org.apache.commons.collections4.IteratorUtils;
import org.apache.commons.lang3.StringUtils;
import org.reactivestreams.Publisher;
import org.slf4j.Logger;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;

import akka.NotUsed;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Status;
import akka.japi.Pair;
import akka.stream.Materializer;
import akka.stream.OverflowStrategy;
import akka.stream.javadsl.AsPublisher;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Keep;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import forms.CreateDashboardItems;
import forms.CreateDashboardNew;
import forms.Item;
import play.data.Form;
import play.data.FormFactory;
import play.libs.F;
import play.libs.Json;
import play.libs.concurrent.HttpExecutionContext;
import play.mvc.Controller;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.Results;
import play.mvc.WebSocket;
import scala.collection.JavaConverters;
import scala.compat.java8.FutureConverters;
import views.html.createDashboardItems;
import views.html.index;




@Singleton
public class Application extends Controller {


    private Logger logger = org.slf4j.LoggerFactory.getLogger("controllers.Application");

    public static final String SESSION_DASHBOARD_ITEMS = "items";
    public static final String SESSION_DASHBOARD_ITEMS_ITEMS = "items";

    public static final String SESSION_DASHBOARD_NAME = "name";
    public static final String SESSION_DASHBOARD_DESCRIPTION = "description";
    public static final String SESSION_DASHBOARD_TYPE = "type";

    public static long TIMEOUT_MILLIS = 100000;

    //private final ActorRef dashboardView;
    private Materializer materializer;
    private ActorSystem actorSystem;
    private FormFactory formFactory;

    private ObjectMapper objectMapper = new ObjectMapper();

    @Inject
    HttpExecutionContext ec;


    @Inject
    public Application(ActorSystem actorSystem,
                       Materializer materializer,
                       FormFactory formFactory) {
        this.materializer = materializer;
        this.actorSystem = actorSystem;
        this.formFactory = formFactory;

        //scodashActor = actorSystem.actorOf(Scodash.props(), Scodash.Name());
        //dashboardView = actorSystem.actorOf(DashboardView.props(), DashboardView.Name());
        //actorSystem.actorOf(DashboardViewBuilder.props(), DashboardViewBuilder.Name());
    }

//    public Result showNewDashboard() {
//
//        Form<CreateDashboardNew> createDashboard1Form = formFactory.form(CreateDashboardNew.class);
//        Form<CreateDashboardNew> filledForm = createDashboard1Form.fill(
//                new CreateDashboardNew(
//                        session(SESSION_DASHBOARD_NAME),
//                        session(SESSION_DASHBOARD_DESCRIPTION),
//                        session(SESSION_DASHBOARD_TYPE)
//                ));
//        return ok(views.html.createDashboardNew.render(filledForm));
//    }
//
//    public Result processNewDashboard() {
//        CreateDashboardNew createDashboardNewForm = formFactory.form(CreateDashboardNew.class).bindFromRequest(request()).get();
//        session(SESSION_DASHBOARD_NAME, createDashboardNewForm.getName());
//        session(SESSION_DASHBOARD_DESCRIPTION, createDashboardNewForm.getDescription());
//        session(SESSION_DASHBOARD_TYPE, createDashboardNewForm.getType());
//        final String sessionItems = session(SESSION_DASHBOARD_ITEMS);
//        Form<CreateDashboardItems> createDashboard2Form = sessionItems != null ? formFactory.form(CreateDashboardItems.class).bind(Json.parse(sessionItems)) : formFactory.form(CreateDashboardItems.class);
//        return ok(createDashboardItems.render(createDashboard2Form));
//
//    }

    public Result addItem() {
        Item addItem = formFactory.form(Item.class).bindFromRequest(request()).get();
        ArrayNode itemsArray;

        if (session(SESSION_DASHBOARD_ITEMS) == null) {
            itemsArray = Json.newObject().putArray(SESSION_DASHBOARD_ITEMS_ITEMS).add(addItem.getItemName());
        } else {
            itemsArray = (ArrayNode)Json.parse(session(SESSION_DASHBOARD_ITEMS)).get(SESSION_DASHBOARD_ITEMS_ITEMS);
            itemsArray.add(addItem.getItemName());
        }


        itemsArray = Json.newArray().addAll(IteratorUtils.toList(itemsArray.elements()).stream().filter(node -> StringUtils.isNotBlank(node.asText())).distinct().collect(Collectors.toList()));
        JsonNode jsonItems = Json.newObject().set(SESSION_DASHBOARD_ITEMS, itemsArray);
        session(SESSION_DASHBOARD_ITEMS, jsonItems.toString());

        return showDashboardItems();
    }

    public Result showDashboardItems() {
        Form<CreateDashboardItems> createDashboard2Form = formFactory.form(CreateDashboardItems.class);
        try {
            Form<CreateDashboardItems> createDashboard2FormFilled = createDashboard2Form.fill(objectMapper.readValue(session(SESSION_DASHBOARD_ITEMS), CreateDashboardItems.class));
            //return ok(createDashboardItems.render(createDashboard2FormFilled));
            return ok(createDashboardItems.render(null));
        } catch (IOException e) {
            logger.error(e.getMessage(), e);
            //return ok(createDashboardItems.render(createDashboard2Form));
            return ok(createDashboardItems.render(null));
        }
    }

    public Result removeItem() {
        Item removeItem = formFactory.form(Item.class).bindFromRequest(request()).get();
        ArrayNode itemsArray = (ArrayNode)Json.parse(session(SESSION_DASHBOARD_ITEMS)).get(SESSION_DASHBOARD_ITEMS);

        List updateItems = IteratorUtils.toList(itemsArray.elements()).stream().filter(item -> !removeItem.getItemName().equals(item.asText())).collect(Collectors.toList());
        JsonNode jsonItems = Json.newObject().set(SESSION_DASHBOARD_ITEMS, Json.newArray().addAll(updateItems));
        session(SESSION_DASHBOARD_ITEMS, jsonItems.toString());

        return showDashboardItems();

    }



//    public Result processDashboardOwner() {
//        Forms.CreateDashboardOwner createDashboard3Form = formFactory.form(Forms.CreateDashboardOwner.class).bindFromRequest(request()).get();
//        session(SESSION_DASHBOARD_OWNER_NAME, createDashboard3Form.ownerName());
//        session(SESSION_DASHBOARD_OWNER_EMAIL, createDashboard3Form.ownerEmail());
//
//        ArrayNode itemsNodes = (ArrayNode) Json.parse(session(SESSION_DASHBOARD_ITEMS)).get(SESSION_DASHBOARD_ITEMS_ITEMS);
//        final Map<String, ItemFO> items = IteratorUtils.toList(itemsNodes.elements()).stream().map(node -> node.asText()).collect(Collectors.toMap(item -> item, item -> new ItemFO(item.toString())));
//
//        try {
//
//            FullResult fr = (FullResult) FutureConverters.toJava(
//                    ask(scodashActor, new Scodash$Command$CreateNewDashboard(session(SESSION_DASHBOARD_NAME),
//                            session(SESSION_DASHBOARD_DESCRIPTION),
//                            session(SESSION_DASHBOARD_TYPE),
//                            JavaConverters.mapAsScalaMapConverter(items).asScala(),
//                            session(SESSION_DASHBOARD_OWNER_NAME),
//                            session(SESSION_DASHBOARD_OWNER_EMAIL)), TIMEOUT_MILLIS)).toCompletableFuture().get();
//            DashboardFO dashboardFO = (DashboardFO) fr.toOption().get();
//
//            forms.Forms.CreatedDashboard createdDashboard = new forms.Forms.CreatedDashboard(dashboardFO.name(), dashboardFO.writeHash(), dashboardFO.readonlyHash());
//
//            Form<Forms.CreatedDashboard> createdDashboardForm = formFactory.form(Forms.CreatedDashboard.class).fill(createdDashboard);
//            return ok(views.html.createdDashboard.render(createdDashboardForm));
//        } catch (Exception e) {
//            return internalServerError();
//        }
//    }

    public WebSocket ws(String hash) {
        return WebSocket.Json.acceptOrResult(request -> {
            if (sameOriginCheck(request)) {
                final CompletionStage<Flow<JsonNode, JsonNode, NotUsed>> future = wsFutureFlow(request, hash);
                final CompletionStage<F.Either<Result, Flow<JsonNode, JsonNode, ?>>> stage = future.thenApplyAsync(F.Either::Right);
                return stage.exceptionally(this::logException);
            } else {
                return forbiddenResult();
            }
        });
    }

    private CompletionStage<F.Either<Result, Flow<JsonNode, JsonNode, ?>>> forbiddenResult() {
        final Result forbidden = Results.forbidden("forbidden");
        final F.Either<Result, Flow<JsonNode, JsonNode, ?>> left = F.Either.Left(forbidden);

        return CompletableFuture.completedFuture(left);
    }


    public boolean sameOriginCheck(Http.RequestHeader rh) {
        final String origin = rh.getHeader("Origin");

        if (origin == null) {
            logger.error("originCheck: rejecting request because no Origin header found");
            return false;
        } else if (originMatches(origin)) {
            logger.debug("originCheck: originValue = " + origin);
            return true;
        } else {
            logger.error("originCheck: rejecting request because Origin header value " + origin + " is not in the same origin");
            return false;
        }
    }


    public CompletionStage<Flow<JsonNode, JsonNode, NotUsed>> wsFutureFlow(Http.RequestHeader request, String hash) {
        // create an actor ref source and associated publisher for sink
        final Pair<ActorRef, Publisher<JsonNode>> pair = createWebSocketConnections();
        ActorRef webSocketOut = pair.first();
        Publisher<JsonNode> webSocketIn = pair.second();

        String id = String.valueOf(request._underlyingHeader().id());
        // Create a user actor off the request id and attach it to the source
        final CompletionStage<ActorRef> userActorFuture = createUserActor(id, webSocketOut, hash);

        // Once we have an actor available, create a flow...
        final CompletionStage<Flow<JsonNode, JsonNode, NotUsed>> stage = userActorFuture
                .thenApplyAsync(userActor -> createWebSocketFlow(webSocketIn, userActor));

        return stage;
    }

    public CompletionStage<ActorRef> createUserActor(String id, ActorRef webSocketOut, String hash) {
//        return FutureConverters.toJava(
//                ask(scodashActor, new Scodash$Command$CreateUser(id, webSocketOut), TIMEOUT_MILLIS)
//        ).thenApply(stageObj -> (ActorRef) stageObj);
        return null;
    }

//    public CompletionStage<ActorRef> createDashboardActor(Dashboard dashboard) {
//
//        // Use guice assisted injection to instantiate and configure the child actor.
//        return FutureConverters.toJava(
//                ask(scodashActor, new Scodash.CreateNewDashboard(dashboard), TIMEOUT_MILLIS)
//        ).thenApply(stageObj -> (ActorRef) stageObj);
//    }

    public Flow<JsonNode, JsonNode, NotUsed> createWebSocketFlow(Publisher<JsonNode> webSocketIn, ActorRef userActor) {
        // http://doc.akka.io/docs/akka/current/scala/stream/stream-flows-and-basics.html#stream-materialization
        // http://doc.akka.io/docs/akka/current/scala/stream/stream-integrations.html#integrating-with-actors

        // source is what comes in: browser ws events -> play -> publisher -> userActor
        // sink is what comes out:  userActor -> websocketOut -> play -> browser ws events
        final Sink<JsonNode, NotUsed> sink = Sink.actorRef(userActor, new Status.Success("success"));
        final Source<JsonNode, NotUsed> source = Source.fromPublisher(webSocketIn);
        final Flow<JsonNode, JsonNode, NotUsed> flow = Flow.fromSinkAndSource(sink, source);

        // Unhook the user actor when the websocket flow terminates
        // http://doc.akka.io/docs/akka/current/scala/stream/stages-overview.html#watchTermination
        return flow.watchTermination((ignore, termination) -> {
            termination.whenComplete((done, throwable) -> {
                logger.info("Terminating actor {}", userActor);
                //dashboardActor.tell(new Stock.Unwatch(null), userActor);
                actorSystem.stop(userActor);
            });

            return NotUsed.getInstance();
        });
    }


    public Pair<ActorRef, Publisher<JsonNode>> createWebSocketConnections() {
        // Creates a source to be materialized as an actor reference.

        // Creating a source can be done through various means, but here we want
        // the source exposed as an actor so we can send it messages from other
        // actors.
        final Source<JsonNode, ActorRef> source = Source.actorRef(10, OverflowStrategy.dropTail());

        // Creates a sink to be materialized as a publisher.  Fanout is false as we only want
        // a single subscriber here.
        final Sink<JsonNode, Publisher<JsonNode>> sink = Sink.asPublisher(AsPublisher.WITHOUT_FANOUT);

        // Connect the source and sink into a flow, telling it to keep the materialized values,
        // and then kicks the flow into existence.
        final Pair<ActorRef, Publisher<JsonNode>> pair = source.toMat(sink, Keep.both()).run(materializer);
        return pair;
    }


    public F.Either<Result, Flow<JsonNode, JsonNode, ?>> logException(Throwable throwable) {
        // https://docs.oracle.com/javase/tutorial/java/generics/capture.html
        logger.error("Cannot create websocket", throwable);
        Result result = Results.internalServerError("error");
        return F.Either.Left(result);
    }


    private boolean originMatches(String origin) {
        return origin.contains("localhost:9000") || origin.contains("localhost:19001");
    }

//    public Result index() {
//        return ok(index.render("Your new application is ready."));
//    }

  //  public Result dashboard(String hash) {
//
//        try {
//            DashboardFO dsh = (DashboardFO) FutureConverters.toJava(ask(dashboardView, new DashboardView.FindDashboardByWriteHash(hash), Application.TIMEOUT_MILLIS)).toCompletableFuture().get();
//            return ok(views.html.dashboard.render(dsh));
//        } catch (Exception e) {
//            e.printStackTrace();
//            return internalServerError();
//        }


        //val d = dashboardView ?

//        try {
//
//
////            String hash = "abc";
//            ActorRef dashboardActor = (ActorRef) FutureConverters.toJava(
//                    ask(scodashActor, new ScodashActor.GetDashboardActor(pojo.DashboardId.apply(hash)), Application.TIMEOUT_MILLIS)
//            ).toCompletableFuture().get();
//            Dashboard dsh = (Dashboard)FutureConverters.toJava(
//                    ask(dashboardActor, new Dashboard.GetDashboard(), Application.TIMEOUT_MILLIS)).toCompletableFuture().get();
//            return ok(views.html.dashboard.render(dsh));
//        } catch (Exception e) {
//            e.printStackTrace();
//            return internalServerError();
//        }

  //  }




}

File: app/controllers/Application.scala
package controllers

import akka.NotUsed
import akka.actor.{ActorRef, ActorSystem}
import akka.event.Logging
import akka.pattern.ask
import akka.stream.scaladsl.{Flow, Keep, Sink, Source}
import akka.stream.{Materializer, OverflowStrategy}
import akka.util.Timeout
import com.google.inject.Inject
import com.google.inject.name.Named
import controllers.Dashboard.Event.DashboardCreated
import controllers.Forms.CreateDashboardItems
import controllers.actors.Scodash.Command.CreateNewDashboard
import controllers.actors.{DashboardAccessMode, Scodash}
import org.apache.commons.lang3.StringUtils
import org.joda.time.DateTimeZone
import org.json4s.ext.JodaTimeSerializers
import org.json4s.native.Serialization.write
import org.json4s.native._
import org.json4s.{DefaultFormats, _}
import org.reactivestreams.Publisher
import play.api.data.Form
import play.api.data.Forms._
import play.api.libs.concurrent.Execution.Implicits.defaultContext
import play.api.libs.json.JsValue
import play.api.mvc._

import scala.concurrent.duration._
import scala.concurrent.{Await, Future}

class Application @Inject() (
                              cc: MessagesControllerComponents,
                              @Named(Scodash.Name) scodashActor: ActorRef,
                              @Named(DashboardView.Name) dashboardViewActor: ActorRef,
                              @Named(DashboardViewBuilder.Name) dashboardViewBuilder: ActorRef,
                              system: ActorSystem,
                              implicit val mat: Materializer)
    extends MessagesAbstractController(cc) {

  // Use a direct reference to SLF4J
  private val logger = org.slf4j.LoggerFactory.getLogger("controllers.Application")

  val SESSION_DASHBOARD = "dashboard"

  implicit val timeout: Timeout = 5.seconds
  implicit lazy val formats = DefaultFormats ++ JodaTimeSerializers.all
//  implicit private val ItemWrites = Json.writes[ItemFO]
//  implicit private val DashoboardWrites = Json.writes[DashboardFO]


  def index() = Action {
    Ok(views.html.index());
  }

  val newDashboardForm = Form(
    mapping(
      "name" -> nonEmptyText,
      "description" -> nonEmptyText
    )(Forms.NewDashboard.apply)(Forms.NewDashboard.unapply)
  )

  def showNewDashboard() = Action { implicit request =>
    request.session.get(SESSION_DASHBOARD) match {
      case Some(sessDash) =>
        Ok(views.html.createDashboardNew(newDashboardForm.fill(JsonMethods.parse(sessDash).extract[Forms.NewDashboard])))
      case _ =>
        Ok(views.html.createDashboardNew(newDashboardForm))
    }
  }

  val dashboardItemsForm = Form(
    mapping(
      "items" -> set(nonEmptyText)
    )(Forms.CreateDashboardItems.apply)(Forms.CreateDashboardItems.unapply)
  )

  def processNewDashboard() = Action { implicit request =>
    newDashboardForm.bindFromRequest.fold(
      formWithErrors => {
        BadRequest(views.html.createDashboardNew(formWithErrors))
      },
      dashboardData => {
        var dashboard = new Forms.Dashboard()
        request.session.get(SESSION_DASHBOARD) match {
          case Some(sessDash) => dashboard = JsonMethods.parse(sessDash).extract[Forms.Dashboard]
          case _ =>
        }
        val updatedDashboard = dashboard.updateNameDescStyle(dashboardData.name, dashboardData.description)
        Ok(views.html.createDashboardItems(dashboardItemsForm.fill(new CreateDashboardItems(updatedDashboard)))).withSession(SESSION_DASHBOARD -> write(updatedDashboard))
      }
    )
  }

  def showDashboardItems() = Action { implicit request =>
    request.session.get(SESSION_DASHBOARD) match {
      case Some(sessDash) => Ok(views.html.createDashboardItems(dashboardItemsForm.fill(JsonMethods.parse(sessDash).extract[CreateDashboardItems])))
      case _ => Ok(views.html.createDashboardItems(dashboardItemsForm))
    }
  }

  val itemForm = Form(
    mapping(
      "itemName" -> text
    )(Forms.Item.apply)(Forms.Item.unapply)
  )

  def addItem() = Action { implicit request =>
    itemForm.bindFromRequest.fold(
      formWithErrors => {
        BadRequest(views.html.createDashboardItems(dashboardItemsForm))
      },
      formData => {
        val sessDash = JsonMethods.parse(request.session.get(SESSION_DASHBOARD).get).extract[Forms.Dashboard]
        if (StringUtils.isNotBlank(formData.itemName)) {
          val updatedDashboard = sessDash.updateItems(sessDash.items + formData.itemName)
          Ok(views.html.createDashboardItems(dashboardItemsForm.fill(new Forms.CreateDashboardItems(updatedDashboard)))).withSession(SESSION_DASHBOARD -> write(updatedDashboard))
        } else {
          Ok(views.html.createDashboardItems(dashboardItemsForm.fill(new Forms.CreateDashboardItems(sessDash))))
        }
      }
    )
  }

  def removeItem() = Action { implicit request =>
    itemForm.bindFromRequest.fold(
      formWithErrors => {
        BadRequest(views.html.createDashboardItems(dashboardItemsForm))
      },
      formData => {
        val sessDash = JsonMethods.parse(request.session.get(SESSION_DASHBOARD).get).extract[Forms.Dashboard]
        val updatedDashboard = sessDash.updateItems(sessDash.items - formData.itemName)
        Ok(views.html.createDashboardItems(dashboardItemsForm.fill(new Forms.CreateDashboardItems(updatedDashboard)))).withSession(SESSION_DASHBOARD -> write(updatedDashboard))
      }
    )

  }

  val dashboardOwnerForm = Form(
    mapping(
      "name" -> nonEmptyText,
      "email" -> email,
      "tzOffset" -> nonEmptyText

    )(Forms.DashboardOwner.apply)(Forms.DashboardOwner.unapply)
  )

  def showDashboardOwner() = Action { implicit request =>
    val sessDash = JsonMethods.parse(request.session.get(SESSION_DASHBOARD).get).extract[Forms.Dashboard]
    if (sessDash.items.size > 0) {
      Ok(views.html.createDashboardOwner(dashboardOwnerForm))
    } else {
      Ok(views.html.createDashboardItems(dashboardItemsForm.withGlobalError("At least one item should be defined in dashboard.")))
    }
  }

  def processDashboardOwner() = Action.async { implicit request =>
    dashboardOwnerForm.bindFromRequest.fold(
      formWithErrors => {
        Future(BadRequest(views.html.createDashboardOwner(formWithErrors)))
      },
      ownerData => {
        val sessDash = JsonMethods.parse(request.session.get(SESSION_DASHBOARD).get).extract[Forms.Dashboard]
        (scodashActor ? CreateNewDashboard(
          sessDash.name,
          sessDash.description,
          sessDash.items.zipWithIndex.map { case (name, id) => ItemFO(id, name) } ,
          ownerData.ownerName,
          ownerData.ownerEmail,
          DateTimeZone.forOffsetHours(Integer.valueOf(ownerData.tzOffset)/60)
          )).mapTo[FullResult[DashboardFO]].map {
          r => {
            dashboardViewBuilder ? DashboardCreated(r.value)
            Ok(views.html.createdDashboard(Forms.CreatedDashboard(r.value.name, r.value.writeHash, r.value.readonlyHash))).withNewSession
          }
        }
      }
    )
  }

  def dashboard(hash: String) = Action.async { implicit request =>
    getDashboard(hash).flatMap{
      case Some((dashboard, accessMode)) =>
        Future(Ok(views.html.dashboard(dashboard)))
      case _ =>
        Future(Ok(views.html.noDashboard()))
    }
  }

  def dashboardData(hash: String) = Action.async { implicit request =>
    getDashboard(hash).flatMap{ case Some((dashboard, accessMode)) =>
      val json = write(dashboard)
      Future(Ok(json))
    }
  }

  def dashboardsData(hashesStr: String) = Action.async { implicit request =>
    val hashes = hashesStr.split(",")
    Future.sequence(hashes.map(hash => getDashboard(hash)).toList).flatMap(ds => {
      val json = write(ds.filter(_.isDefined).map(_.get._1))
      Future(Ok(json))
    })
  }

  /**
    * Creates a websocket.  `acceptOrResult` is preferable here because it returns a
    * Future[Flow], which is required internally.
    *
    * @return a fully realized websocket.
    */
  def ws(hash: String): WebSocket = WebSocket.acceptOrResult[JsValue, JsValue] {
    case rh if sameOriginCheck(rh) =>
      wsFutureFlow(rh, hash).map { flow =>
        Right(flow)
      }.recover {
        case e: Exception =>
          logger.error("Cannot create websocket", e)
          val jsError =  play.api.libs.json.Json.obj("error" -> "Cannot create websocket")
          val result = InternalServerError(jsError)
          Left(result)
      }

    case rejected =>
      logger.error(s"Request ${rejected} failed same origin check")
      Future.successful {
        Left(Forbidden("forbidden"))
      }
  }

  /**
    * Checks that the WebSocket comes from the same origin.  This is necessary to protect
    * against Cross-Site WebSocket Hijacking as WebSocket does not implement Same Origin Policy.
    *
    * See https://tools.ietf.org/html/rfc6455#section-1.3 and
    * http://blog.dewhurstsecurity.com/2013/08/30/security-testing-html5-websockets.html
    */
  def sameOriginCheck(rh: RequestHeader): Boolean = {
    rh.headers.get("Origin") match {
      case Some(originValue) if originMatches(originValue) =>
        logger.debug(s"originCheck: originValue = $originValue")
        true

      case Some(badOrigin) =>
        logger.error(s"originCheck: rejecting request because Origin header value $badOrigin is not in the same origin")
        false

      case None =>
        logger.error("originCheck: rejecting request because no Origin header found")
        false
    }
  }

  /**
    * Returns true if the value of the Origin header contains an acceptable value.
    */
  def originMatches(origin: String): Boolean = {
    val host = sys.env("HOST")
    origin.contains(host)
  }

  /**
    * Creates a Future containing a Flow of JsValue in and out.
    */
  def wsFutureFlow(request: RequestHeader, hash: String): Future[Flow[JsValue, JsValue, NotUsed]] = {
    // create an actor ref source and associated publisher for sink
    val (webSocketOut: ActorRef, webSocketIn: Publisher[JsValue]) = createWebSocketConnections()

    // Create a user actor off the request id and attach it to the source
    val userActorFuture = createUserActor(request.id.toString, webSocketOut, hash)

    // Once we have an actor available, create a flow...
    userActorFuture.map { userActor =>
      createWebSocketFlow(webSocketIn, userActor, hash)
    }
  }

  /**
    * Creates a materialized flow for the websocket, exposing the source and sink.
    *
    * @return the materialized input and output of the flow.
    */
  def createWebSocketConnections(): (ActorRef, Publisher[JsValue]) = {

    // Creates a source to be materialized as an actor reference.
    val source: Source[JsValue, ActorRef] = {
      // If you want to log on a flow, you have to use a logging adapter.
      // http://doc.akka.io/docs/akka/2.4.4/scala/logging.html#SLF4J
      val logging = Logging(system.eventStream, logger.getName)

      // Creating a source can be done through various means, but here we want
      // the source exposed as an actor so we can send it messages from other
      // actors.
      Source.actorRef[JsValue](10, OverflowStrategy.dropTail).log("actorRefSource")(logging)
    }

    // Creates a sink to be materialized as a publisher.  Fanout is false as we only want
    // a single subscriber here.
    val sink: Sink[JsValue, Publisher[JsValue]] = Sink.asPublisher(fanout = false)

    // Connect the source and sink into a flow, telling it to keep the materialized values,
    // and then kicks the flow into existence.
    source.toMat(sink)(Keep.both).run()
  }

  /**
    * Creates a flow of events from the websocket to the user actor.
    *
    * When the flow is terminated, the user actor is no longer needed and is stopped.
    *
    * @param userActor   the user actor receiving websocket events.
    * @param webSocketIn the "read" side of the websocket, that publishes JsValue to UserActor.
    * @return a Flow of JsValue in both directions.
    */
  def createWebSocketFlow(webSocketIn: Publisher[JsValue], userActor: ActorRef, hash: String): Flow[JsValue, JsValue, NotUsed] = {
    // http://doc.akka.io/docs/akka/current/scala/stream/stream-flows-and-basics.html#stream-materialization
    // http://doc.akka.io/docs/akka/current/scala/stream/stream-integrations.html#integrating-with-actors

    // source is what comes in: browser ws events -> play -> publisher -> userActor
    // sink is what comes out:  userActor -> websocketOut -> play -> browser ws events
    val flow = {
      val sink = Sink.actorRef(userActor, akka.actor.Status.Success(()))
      val source = Source.fromPublisher(webSocketIn)
      Flow.fromSinkAndSource(sink, source)
    }

    // Unhook the user actor when the websocket flow terminates
    // http://doc.akka.io/docs/akka/current/scala/stream/stages-overview.html#watchTermination
    val flowWatch: Flow[JsValue, JsValue, NotUsed] = flow.watchTermination() { (_, termination) =>
      termination.foreach { done =>
        logger.info(s"Terminating actor $userActor")
        (dashboardViewActor ? Scodash.Command.FindDashboardByWriteHash(hash)).mapTo[FullResult[List[JObject]]].map {
          result => {
            val dashboardFO = result.value.head.extract[DashboardFO]
            (scodashActor ? (Scodash.Command.FindDashboard(dashboardFO.id))).mapTo[ActorRef].map {
              dashboardActor =>
                dashboardActor ! Dashboard.Command.Unwatch
                system.stop(userActor)
            }
          }
        }
      }
      NotUsed
    }

    flowWatch
  }

  /**
    * Creates a user actor with a given name, using the websocket out actor for output.
    *
    * @param userId         the name of the user actor.
    * @param webSocketOut the "write" side of the websocket, that the user actor sends JsValue to.
    * @return a user actor for this ws connection.
    */
  def createUserActor(userId: String, webSocketOut: ActorRef, hash: String): Future[ActorRef] = {
    val userActorFuture = {
      getDashboard(hash).flatMap{case Some((dashboard, accessMode)) =>
        (scodashActor ? Scodash.Command.CreateDashboardUser(userId, webSocketOut, dashboard.id, accessMode)).mapTo[ActorRef]
      }


    }
    userActorFuture



  }

  def getDashboard(hash: String): Future[Option[(DashboardFO, DashboardAccessMode.Value)]] = {

    var writeFut = scodashActor ? Scodash.Command.FindDashboardByWriteHash(hash)
    var readFut = scodashActor ? Scodash.Command.FindDashboardByReadonlyHash(hash)

    var dashboard = resolveDashboard(writeFut, readFut)
    dashboard.map {
      case Some((dashboard, mode)) => {
        logger.info("Found {} dashboard in maps", hash)
        Some((dashboard, mode))
      }
      case None => {
        logger.info("Not found {} dashboard in maps - going to read model", hash)
        writeFut = dashboardViewActor ? Scodash.Command.FindDashboardByWriteHash(hash)
        readFut = dashboardViewActor ? Scodash.Command.FindDashboardByReadonlyHash(hash)
        Await.result(resolveDashboard(writeFut, readFut), 10 seconds)
      }
    }
  }

  private def resolveDashboard(writeFut: Future[Any], readFut: Future[Any]): Future[Option[(DashboardFO, DashboardAccessMode.Value)]] = {
    for {
      writeDash <- writeFut
      readDash <- readFut
    } yield {
      val maybeReadDashboard = readDash match {
        case readRes: DashboardFO =>
          Some(readRes.removeWriteHash, DashboardAccessMode.READONLY)
        case readRes: FullResult[_] =>
          readRes.value match {
            case List(item) =>
              item match {
                case item: DashboardFO => Some(item.removeWriteHash, DashboardAccessMode.READONLY)
                case jsObject: JObject => Some(jsObject.extract[DashboardFO].removeWriteHash, DashboardAccessMode.READONLY)
              }
            case _ => None
          }
        case _ => None
      }
      writeDash match {
        case writeRes: DashboardFO =>
          Some(writeRes.removeReadOnlyHash, DashboardAccessMode.WRITE)
        case writeRes: FullResult[_] =>
          writeRes.value match {
            case List(item) =>
              item match {
                case dashboard: DashboardFO => Some(dashboard.removeReadOnlyHash, DashboardAccessMode.WRITE)
                case jsObject: JObject => Some(jsObject.extract[DashboardFO].removeReadOnlyHash, DashboardAccessMode.WRITE)
              }
            case _ => None
          }
        case _ => None
      }
    }
  }
}